src/VisualStudio/Core/Def/Implementation/Diagnostics/SolutionCrawlerEventHandlerWithDelay.cs

```diff
@@ -0,0 +1,124 @@
+                _isTimerRunning = false;
+
+                // no event unsubscription since it will remain alive until VS shutdown
+                reporter.ProgressChanged += OnSolutionCrawlerProgressChanged;
+            }
+
+            /// <summary>
+            /// Retrieve and throttle solution crawler events to be sent to the progress reporter UI.
+            /// 
+            /// there is no concurrent call to this method since ISolutionCrawlerProgressReporter will serialize all
+            /// events to preserve event ordering
+            /// </summary>
+            /// <param name="progressData"></param>
+            public void OnSolutionCrawlerProgressChanged(object sender, ProgressData progressData)
+            {
+                lock (_lock)
+                {
+                    _progressData = progressData;
+
+                    // The publish event timer is already running. If this is a stop event make sure the timer does not get extended.
+                    // Otherwise, just update the progress data and continue.
+                    if (_isTimerRunning)
+                    {
+                        if (progressData.Status == ProgressStatus.Stopped)
+                        {
+                            _isTimerRunning = false;
+                        }
+                        else
+                        {
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        // Event publishing is not running, so trigger new timer.
+                        _isTimerRunning = true;
+                        RunOnTimer();
+                    }
+                }
+            }
+
+            private void InvokeEvent()
+            {
+                _invokeEventTask = _invokeEventTask.SafeContinueWith(_ =>
+                {
+                    // No need to publish the exact same event over and over.
+                    if (!Equals(_progressData, _lastPublishedProgressData))
+                    {
+                        UpdateProgress?.Invoke(this, _progressData);
+                        _lastPublishedProgressData = _progressData;
+                    }
+                }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+            }
+
+            private void RunOnTimer()
+            {
+                lock (_lock)
+                {
+                    InvokeEvent();
+
+                    // Only continue the timer if there has not been a stop event.
+                    if (_isTimerRunning)
+                    {
+                        _intervalTask = _intervalTask.ContinueWithAfterDelay(() =>
+                        {
+                            RunOnTimer();
+                        }, CancellationToken.None, _publishInterval, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+                    }
+                }
```

Comment:
i completely couldn't understand what was going on here.  And doing things like invoking an event under a lock seem super scary.